Subject: qeth: set new mac even if old mac is gone
From: Ursula Braun <ursula.braun@de.ibm.com>
Patch-mainline: v3.7-rc6
Commit-id: 7702745b15128e5f0659693736a864e35be1c807
References: bnc#789010,LTC#86643

Symptom:      qeth slaves of bonding in active_backup mode and
              fail_over_mac=2 show same mac address after failover.
Problem:      If the set_mac_address() function of qeth is invoked,
              qeth deletes the old mac address first on OSA. Only 
              if deletion returns successfully the new mac address
              is set on OSA. Deletion may return with a return value
              "MAC not found on OSA". In this case qeth should 
              continue setting the new mac address.
              When the OSA cable is pulled, OSA forgets any set 
              mac address. If the OSA network interface acts as a 
              slave to a bonding master interface, bonding can
              invoke the set_mac_address function for failover
              purposes and depends on successful setting of the new
              mac address even though the old mac address could no
              longer be deleted.
Solution:     Continue setting the new mac address even if deletion
              of old mac address fails with "MAC not found on OSA"
Reproduction: Setup bonding interface in active_backup mode and
              fail_over_mac=2 with 2 OSA-cards as slaves and enforce
              failover by pullling the cable of the active OSA slave.

Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/net/qeth_l2_main.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/s390/net/qeth_l2_main.c
+++ b/drivers/s390/net/qeth_l2_main.c
@@ -676,7 +676,7 @@ static int qeth_l2_set_mac_address(struc
 		return -ERESTARTSYS;
 	}
 	rc = qeth_l2_send_delmac(card, &card->dev->dev_addr[0]);
-	if (!rc)
+	if (!rc || (rc == IPA_RC_L2_MAC_NOT_FOUND))
 		rc = qeth_l2_send_setmac(card, addr->sa_data);
 	return rc;
 }
